rust/lance-index/src/scalar/ngram.rs

```diff
@@ -434,127 +593,452 @@ impl NGramIndexBuilder {
-                    builder.process_batch(&batch);
+                while let Some(partition) = recv.recv().await {
+                    builder.process_batch(partition).await?;
                 }
-                builder
+                Result::Ok(builder)
             });
             builders.push(future);
         }
 
-        let mut idx = 0;
-        while let Some(batch) = data.try_next().await? {
-            senders[idx % num_shards].send(batch).await.unwrap();
-            idx += 1;
+        let mut partitions_stream = data
+            .and_then(|batch| {
+                let tokenizer = self.tokenizer.clone();
+                std::future::ready(Ok(tokio::task::spawn(async move {
+                    Ok(Self::tokenize_and_partition(&tokenizer, batch, num_workers))
+                })
+                .map(|res| res.unwrap())))
+            })
+            .try_buffer_unordered(*DEFAULT_TOKENIZE_PARALLELISM);
+
+        while let Some(partitions) = partitions_stream.try_next().await? {
+            for (part_idx, partition) in partitions.into_iter().enumerate() {
+                senders[part_idx].send(partition).await.unwrap();
+            }
         }
 
         std::mem::drop(senders);
         let builders = futures::future::try_join_all(builders).await?;
+
+        // Final flush is serialized.  If we kick this off in parallel it can
+        // use a lot of memory.
+
+        let mut to_spill = Vec::with_capacity(builders.len());
+
         for builder in builders {
-            self.merge(builder);
+            let mut builder = builder?;
+            let state = builder.state.take();
+            if builder.flush(state).await? {
+                to_spill.push(builder.worker_number);
+            }
         }
 
+        Ok(to_spill)
+    }
+
+    async fn write(
+        &mut self,
+        mut writer: Box<dyn IndexWriter>,
+        state: NGramIndexSpillState,
+    ) -> Result<()> {
+        writer.write_record_batch(state.try_into_batch()?).await?;
+        writer.finish().await?;
+
         Ok(())
     }
 
-    fn merge(&mut self, mut other: Self) {
-        for (token, new_token_id) in other.tokens_map {
-            if let Some(token_id) = self.tokens_map.get(&token) {
-                self.bitmaps[*token_id as usize] |=
-                    std::mem::take(&mut other.bitmaps[new_token_id as usize]);
+    async fn stream_spill(
+        &self,
+        id: usize,
+    ) -> Result<impl Stream<Item = Result<NGramIndexSpillState>>> {
+        let reader = self
+            .spill_store
+            .open_index_file(&Self::spill_filename(id))
+            .await?;
+
+        let num_rows = reader.num_rows();
+
+        Ok(stream::try_unfold(0, move |offset| {
+            let reader = reader.clone();
+            async move {
+                // These are rediculously small batches but, in the worst case scenario, each row could
+                // be massive (up to 128MB per row at 1B rows) and we end up breaking memory
```

Comment:
"ridiculously". Also it would be helpful to mention in this comment that those numbers come from LANCE_NGRAM_* environment configs.